### PR TITLE
add data parallel for transformer

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
## Description ##
Add a python multi-thread forward utility, applied to transformer. Throughput:
- previous: 110K/s
- now: 160K/s

@szhengac @szha @bikestra @hhexiy @leezu 

Note that the Gluon blocks still need to be initialized using a single thread (by forwarding with a batch)... The interface is not ideal but at least works. 

### Essentials ###
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: scripts/machine_translation/transformer.py [hunk 1]
```diff
@@ -988,3 +989,28 @@ def get_model(name, dataset_name='WMT2014', **kwargs):
...
+        rescaled_ls = (ls * (tgt_seq.shape[1] - 1)) / batch_size / 100.0
+        return ls, rescaled_ls
+
+    def backward(self, loss):
+        loss.backward()
+
+    def forward_backward(self, x):
+        sequence, batch_size = x
+        ls, rescaled_ls = self.forward(*sequence, batch_size)
+        self.backward(ls)
```
Comment (szhengac) [inline scripts/machine_translation/transformer.py]: In the original code, the rescaled_ls is used to perform the backpropagation.

**Comment by szhengac** [comment]
So in the original gluon api, using a simple for loop does not really achieve parallelization?

**Comment by leezu** [comment]
@szhengac It does, but as the Transformer model is somewhat complex, each `.forward()` is too expensive and blocks the main Python thread too long to fully utilize multiple GPUs. So the first GPU may finish before you push the Operator to the last GPU. Haibin has more details

File: scripts/machine_translation/transformer.py (hunk 1 above)
Comment (eric-haibin-lin) [reply scripts/machine_translation/transformer.py]: My bad. Will fix it

**Comment by eric-haibin-lin** [comment]
@szhengac Yes calling block.forward with single thread is not fast enough to keep the GPU busy, specially when a block is not hybridized and involves many small ndarray imperative calls

File: gluonnlp/utils/parallel.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+                for x in gluon.utils.split_and_load(batch, ctx):
+                    parallel.put(x)
+                losses = [parallel.get() for _ in ctx]
+            trainer.step()
+    """
+    def forward_backward(self, x):
+        """ Forward and backward computation. """
+        raise NotImplementedError()
+
+class Parallel:
```
Comment (leezu) [inline gluonnlp/utils/parallel.py]: Should inherit from `object` for Py2 support?

File: gluonnlp/utils/parallel.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+"""Utility functions for parallel processing."""
+import threading
+try:
+    import Queue as queue
+except ImportError:
+    import queue
+
+__all__ = ['Parallelizable', 'Parallel']
+
+class Parallelizable:
```
Comment (leezu) [inline gluonnlp/utils/parallel.py]: Should inherit from object for Py2 support?

File: gluonnlp/utils/__init__.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# pylint: disable=wildcard-import, arguments-differ
+r"""Module for utility functions.
```
Comment (leezu) [inline gluonnlp/utils/__init__.py]: Why a raw string?

Comment (eric-haibin-lin) [reply gluonnlp/utils/__init__.py]: copied from some other file. Does it make any difference?

Comment (leezu) [reply gluonnlp/utils/__init__.py]: I don't think so. Just surprised to see it

File: scripts/machine_translation/transformer.py [hunk 5]
```diff
@@ -988,3 +989,28 @@ def get_model(name, dataset_name='WMT2014', **kwargs):
...
+        self._label_smoothing = label_smoothing
+        self._loss = loss_function
+
+    def forward(self, src_seq, tgt_seq, src_valid_length, tgt_valid_length, batch_size):
+        with mx.autograd.record():
+            out, _ = self._model(src_seq, tgt_seq[:, :-1],
+                                 src_valid_length, tgt_valid_length - 1)
+            smoothed_label = self._label_smoothing(tgt_seq[:, 1:])
+            ls = self._loss(out, smoothed_label, tgt_valid_length - 1).sum()
+            ls = (ls * (tgt_seq.shape[1] - 1)) / batch_size / 100.0
```
Comment (szhengac) [inline scripts/machine_translation/transformer.py]: 100.0 can be taken as a keyword arg.

**Comment by szhengac** [comment]
Btw, during my testing, I found that the threads are not stopped and the process is hanging there.

File: scripts/machine_translation/transformer.py (hunk 5 above)
Comment (eric-haibin-lin) [reply scripts/machine_translation/transformer.py]: Added. Thanks

**Comment by mli** [comment]
Job PR-387/10 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-387/10/index.html

**Comment by eric-haibin-lin** [comment]
I've added unit test. @leezu @szhengac do you mind doing one last round of review?

**Comment by szhengac** [comment]
Have you solved the hanging problem?

**Comment by eric-haibin-lin** [comment]
@szhengac I fixed extra reference that causes hanging in https://github.com/dmlc/gluon-nlp/pull/387/commits/66e2348c7b9b52a1aa52c9bb86eb3290a6baba0e

File: src/gluonnlp/utils/parallel.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+import threading
+try:
+    import Queue as queue
+except ImportError:
+    import queue
+
+__all__ = ['Parallelizable', 'Parallel']
+
+class Parallelizable(object):
+    """ Base class for parallelizable unit of work, which can be invoked by `Parallel`.
```
Comment (leezu) [inline src/gluonnlp/utils/parallel.py]: I believe docstring should not start with whitespace, at least according to all examples in PEP 257 https://www.python.org/dev/peps/pep-0257/

File: src/gluonnlp/model/transformer.py [hunk 7]
```diff
@@ -1208,3 +1209,35 @@ def transformer_en_de_512(dataset_name=None, src_vocab=None, tgt_vocab=None, pre
                                   predefined_args['share_embed'], predefined_args['embed_size'],
                                   predefined_args['tie_weights'],
                                   predefined_args['embed_initializer'], pretrained, ctx, root)
+
+
+class ParallelTransformer(Parallelizable):
+    """ Data parallel transformer.
```
Comment (leezu) [inline src/gluonnlp/model/transformer.py]: Also here, whitespace, see PEP 257 https://www.python.org/dev/peps/pep-0257/

File: src/gluonnlp/model/transformer.py [hunk 8]
```diff
@@ -1208,3 +1209,35 @@ def transformer_en_de_512(dataset_name=None, src_vocab=None, tgt_vocab=None, pre
...
+    rescale_loss : float
+        The scale to which the loss is rescaled to avoid gradient explosion.
+    """
+    def __init__(self, model, label_smoothing, loss_function, rescale_loss):
+        self._model = model
+        self._label_smoothing = label_smoothing
+        self._loss = loss_function
+        self._rescale_loss = rescale_loss
+
+    def forward_backward(self, x):
```
Comment (leezu) [inline src/gluonnlp/model/transformer.py]: Do you know why pylint does not complain about missing docstring here?

File: src/gluonnlp/utils/__init__.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# pylint: disable=wildcard-import, arguments-differ
+r"""Module for utility functions.
+"""
```
Comment (leezu) [inline src/gluonnlp/utils/__init__.py]: Let's follow PEP 257 https://www.python.org/dev/peps/pep-0257/

File: src/gluonnlp/utils/parallel.py [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+            self._num_serial -= 1
+            out = self._parallizable.forward_backward(x)
+            self._out_queue.put(out)
+        else:
+            self._in_queue.put(x)
+
+    def get(self):
+        """ Get an output of previous `parallizable.forward_backward` calls.
+        This method blocks if none of previous `parallizable.forward_backward`
+        calls have return any result. """
```
Comment (leezu) [inline src/gluonnlp/utils/parallel.py]: PEP 257 https://www.python.org/dev/peps/pep-0257/

File: src/gluonnlp/utils/parallel.py [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+        Parallelizable net whose `forward` and `backward` methods are invoked
+        by multiple worker threads.
+    serial_init : bool, default True
+        Execute the first `num_workers` inputs in main thread, so that the `Block`
+        used in `parallizable` is initialized serially. Initialize a `Block` with
+        multiple threads may cause unexpected behavior.
+    """
+
+    class _StopSignal(object):
+        """ Internal class to signal stop. """
```
Comment (leezu) [inline src/gluonnlp/utils/parallel.py]: PEP 257 https://www.python.org/dev/peps/pep-0257/

File: src/gluonnlp/utils/parallel.py [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+
+        arg = (self._in_queue, self._out_queue, self._parallizable)
+        for _ in range(num_workers):
+            thread = threading.Thread(target=_worker, args=arg)
+            self._threads.append(thread)
+            thread.start()
+
+    def put(self, x):
+        """ Assign input `x` to an available worker and invoke
+        `parallizable.forward_backward` with x. """
```
Comment (leezu) [inline src/gluonnlp/utils/parallel.py]: PEP 257 https://www.python.org/dev/peps/pep-0257/

File: tests/unittest/test_utils.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
+import pytest
+import numpy as np
+import mxnet as mx
+import gluonnlp as nlp
+
+
+def test_parallel():
```
Comment (leezu) [inline tests/unittest/test_utils.py]: Please mark the test following https://github.com/dmlc/gluon-nlp/pull/417

File: src/gluonnlp/model/transformer.py (hunk 8 above)
Comment (eric-haibin-lin) [reply src/gluonnlp/model/transformer.py]: Maybe due to https://github.com/dmlc/gluon-nlp/blob/master/.pylintrc#L327-L329 ?

File: tests/unittest/test_utils.py (hunk 13 above)
Comment (eric-haibin-lin) [reply tests/unittest/test_utils.py]: Marked

File: src/gluonnlp/utils/parallel.py (hunk 10 above)
Comment (eric-haibin-lin) [reply src/gluonnlp/utils/parallel.py]: Fixed

File: src/gluonnlp/utils/parallel.py (hunk 12 above)
Comment (eric-haibin-lin) [reply src/gluonnlp/utils/parallel.py]: Fixed

**Comment by mli** [comment]
Job PR-387/17 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-387/17/index.html

**Pull request merged**